README.md
# Ftpd
[![Gem Version](https://badge.fury.io/rb/ftpd.png)](http://badge.fury.io/rb/ftpd)
[![Dependency Status](https://gemnasium.com/wconrad/ftpd.svg)](https://gemnasium.com/wconrad/ftpd)
[![Build Status](https://travis-ci.org/wconrad/ftpd.png)](https://travis-ci.org/wconrad/ftpd)
[![Code Climate](https://codeclimate.com/github/wconrad/ftpd.png)](https://codeclimate.com/github/wconrad/ftpd)

ftpd is a pure Ruby FTP server library.  It supports implicit and
explicit TLS, IPV6, passive and active mode, and is unconditionally
compliant per [RFC-1123][1].  It can be used as part of a test fixture
or embedded in a program.

## A note about this README

This readme contains [Yardoc](http://yardoc.org/) markup for links to
the API docs; those links don't display properly on github.  You'll
find a properly rendered version [on
rubydoc.info](http://rubydoc.info/gems/ftpd)

## The state of this library

Ftpd has been used for many years to test FTP clients, and is stable
and reliable for that purpose.  However, it was not originally
intended to be part of a publically accessible FTP server.  I would be
cautious in using it in an untrusted environment due to the
probability that it contains critical flaws (or even security
vulnarabilities) that have not been discovered in its use as a test
harness.

## Hello World

This is examples/hello_world.rb, a bare minimum FTP server.  It allows
any user/password, and serves files in a temporary directory.  It
binds to an ephemeral port on the local interface:

```ruby
require 'ftpd'
require 'tmpdir'

class Driver

  def initialize(temp_dir)
    @temp_dir = temp_dir
  end

  def authenticate(user, password)
    true
  end

  def file_system(user)
    Ftpd::DiskFileSystem.new(@temp_dir)
  end

end

Dir.mktmpdir do |temp_dir|
  driver = Driver.new(temp_dir)
  server = Ftpd::FtpServer.new(driver)
  server.start
  puts "Server listening on port #{server.bound_port}"
  gets
end
```

A more full-featured example that allows TLS and takes options is in
examples/example.rb

## Driver

Ftpd's dynamic behavior such as authentication and file retrieval is
controlled by a driver that you supply.  The Driver class in the
"hello world" example above shows a rudimentary driver.  Ftpd calls
the authenticate method to decide who can log in.  Once someone is
logged on, it calls the file_system method to obtain a file system
driver for that user.

There is no base class for a driver.  Any object that quacks like a
driver will do.  Here are the methods your driver needs:

* {Example::Driver#authenticate authenticate}
* {Example::Driver#file_system file_system}

## File System

The file system object that the driver supplies to Ftpd is Ftpd's
gateway to the logical file system.  Ftpd doesn't know or care whether
it's serving files from disk, memory, or any other means.

The file system can be very minimal.  If the file system is missing
certain methods, the server simply disables the commands which need
that method.  For example, if there is no write method, then STOR is
not supported and causes a "502 Command not implemented" response to
the client.

The canonical and commented example of an Ftpd file system is
{Ftpd::DiskFileSystem}.  You can use it as a template for creating
your own, and its comments are the official specification for an Ftpd
file system.

Here are the methods a file system may expose:

* {Ftpd::DiskFileSystem::Accessors#accessible? accessible?}
* {Ftpd::DiskFileSystem::Accessors#directory? directory?}
* {Ftpd::DiskFileSystem::Accessors#exists? exists?}
* {Ftpd::DiskFileSystem::Append#append append}
* {Ftpd::DiskFileSystem::Delete#delete delete}
* {Ftpd::DiskFileSystem::List#dir dir}
* {Ftpd::DiskFileSystem::List#file_info file_info}
* {Ftpd::DiskFileSystem::Mkdir#mkdir mkdir}
* {Ftpd::DiskFileSystem::Read#read read}
* {Ftpd::DiskFileSystem::Rename#rename rename}
* {Ftpd::DiskFileSystem::Rmdir#rmdir rmdir}
* {Ftpd::DiskFileSystem::Write#write write}

### DiskFileSystem

Ftpd includes a disk based file system:

```ruby
class Driver

  ...

  def file_system(user)
    Ftpd::DiskFileSystem.new('/var/lib/ftp')
  end

end
```

**Warning**: The DiskFileSystem allows file and directory modification
including writing, renaming, deleting, etc.  If you want a read-only
file system, then use {Ftpd::ReadOnlyDiskFileSystem} instead.

The DiskFileSystem is composed out of modules:

* {Ftpd::DiskFileSystem::Base Base} - You will need this
* {Ftpd::DiskFileSystem::Append Append} - File appending
* {Ftpd::DiskFileSystem::Delete Delete} - File deletion
* {Ftpd::DiskFileSystem::List List} - Directory listing
* {Ftpd::DiskFileSystem::Mkdir Mkdir} - Directory creation
* {Ftpd::DiskFileSystem::Read Read} - File reading
* {Ftpd::DiskFileSystem::Rename Rename} - File renaming
* {Ftpd::DiskFileSystem::Rmdir Rmdir} - Directory removal
* {Ftpd::DiskFileSystem::Write Write} - File writing

You can use these modules to create a custom disk file system that
allows only the operations you want, or which mixes the predefined
modules with your customizations, as in this silly example that allows
uploads but then throws them away.

```ruby
class BlackHole
  def write(ftp_path, contents)
  end
end

class CustomDiskFileSystem
  include DiskFileSystem::Base
  include DiskFileSystem::Read
  include BlackHole
end
```

## Configuration

Configuration is done via accessors on {Ftpd::FtpServer}.  For
example, to set the session timeout to 10 minutes:

```ruby
server = Ftpd::FtpServer.new(driver)
server.session_timeout = 10 * 60
server.start
```

You can set any of these attributes before starting the server:

* {Ftpd::FtpServer#nat_ip}
* {Ftpd::FtpServer#passive_ports}
* {Ftpd::FtpServer#allow_low_data_ports}
* {Ftpd::FtpServer#auth_level}
* {Ftpd::FtpServer#exception_handler}
* {Ftpd::FtpServer#failed_login_delay}
* {Ftpd::FtpServer#list_formatter}
* {Ftpd::FtpServer#log}
* {Ftpd::FtpServer#max_connections_per_ip}
* {Ftpd::FtpServer#max_connections}
* {Ftpd::FtpServer#max_failed_logins}
* {Ftpd::FtpServer#response_delay}
* {Ftpd::FtpServer#server_name}
* {Ftpd::FtpServer#server_version}
* {Ftpd::FtpServer#session_timeout}
* {Ftpd::Server#interface}
* {Ftpd::Server#port}
* {Ftpd::TlsServer#certfile_path}
* {Ftpd::TlsServer#tls}

### LIST output format

By default, the LIST command uses Unix "ls -l" formatting:

    -rw-r--r-- 1 user     group        1234 Mar  3 08:38 foo

An alternative to "ls -l" formatting is [Easily Parsed LIST format
(EPLF)](http://cr.yp.to/ftp/list/eplf.html) format:

    +r,s1234,m1362325080\tfoo

to configure Ftpd for EPLF formatting:

    ftp_server.list_formatter = Ftpd::ListFormat::Eplf

To create your own custom formatter, create a class with these
methods:

* {Ftpd::ListFormat::Ls#initialize initialize}
* {Ftpd::ListFormat::Ls#to_s to_s}

And register your class with the ftp_server before starting it:

    ftp_server.list_formatter = MyListFormatter

### Logging

Ftpd can write to an instance of
{http://www.ruby-doc.org/stdlib-2.0.0/libdoc/logger/rdoc/Logger.html
Logger} that you provide.  To log to standard out:

    server.log = Logger.new($stdout)

To log to a file:

    server.log = Logger.new('/tmp/ftpd.log')

## Standards Compliance

* Unconditionally compliant per [RFC-1123][1] (Requirements for
  Internet Hosts).

* Implements all of the security recommendations in
  [RFC-2577](http://tools.ietf.org/rfc/rfc2577.txt) (FTP Security
  Considerations).

* Implements [RFC-2389](http://tools.ietf.org/rfc/rfc2389.txt)
  (Feature negotiation mechanism for the File Transfer Protocol)

* Implements [RFC-2428](http://tools.ietf.org/rfc/rfc2428.txt) (FTP
  Extensions for IPv6 and NATs)

* Implements enough of
  [RFC-4217](http://tools.ietf.org/rfc/rfc4217.txt) (Securing FTP with
  TLS) to get by.

See [RFC Compliance](doc/rfc-compliance.md) for details

## Ruby Compatability

The tests pass with these Rubies:

* ruby-1.9.3
* ruby-2.0
* ruby-2.1
* ruby-2.2
* ruby-2.3
* ruby-2.4 (but see below)

For Ruby 1.8, use an ftpd version before 0.8.  In your Gemfile:

    gem 'ftpd', '<0.8'

This gem runs fine in Ruby 2.4, but the tests that use TLS are skipped
in Ruby 2.4.  That is because the double_bag_ftps gem that the tests
use for TLS does not work in Ruby 2.4.

## OS compatability

## *nix

Ftpd runs on:

* Linux
* OSX

## Windows

The master branch of ftpd does not currently run on Windows.  There is
an experimental branch for Windows which contains several changes that
ought to make ftpd work on Windows, but they need testing.  To try
that branch, use this line in your Gemfile:

    gem 'ftpd', github: 'wconrad/ftpd', branch: 'windows'

Does it work for you?  Is it busted?  Please report your experience
[here](https://github.com/wconrad/ftpd/issues/3).

## Versioning

This library uses [Semantic Versioning](http://semver.org/).  Ftpd
promises not to make breaking changes to its API without bumping the
major version.

## Development

### Tests

To run the cucumber (functional) tests:

    $ rake test:features

To run the rspec (unit) tests:

    $ rake test:spec

To run all tests:

    $ rake test

or just:

    $ rake

To force features to write the server log to stdout:

    $ FTPD_DEBUG=1 rake test:features

### Example

The stand-alone example is good for manually testing Ftpd with any FTP
client.  To run the stand-alone example:

    $ examples/example.rb

The example prints its port, username and password to the console.
You can connect to the stand-alone example with any FTP client.

example.rb has many options.  To see them:

    $ examples/example.rb -h

### Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Origin

I created ftpd to support the test framework I wrote for Databill,
LLC, which has given its kind permission to donate it to the
community.

## Whoami

Wayne Conrad <wconrad@yagni.com>

## Credits

Thanks to Databill, LLC, which supported the creation of this library,
and granted permission to donate it to the community.

### Contributors

Among those who have improved ftpd are:

* Alfonso Cora
* André Aizim Kelmanson
* Bjoern B. Dorra
* Joshua Rutherford
* Larry. W. Cashdollar
* Michael de Silva
* Mike Ragalie
* cransom

If I've forgotten to add you, please remind me, or submit a merge
request.

Thank you!

## See also

* [Changelog](Changelog.md)
* [RFC compliance](doc/rfc-compliance.md)
* [References](doc/references.md)
* [Benchmarks](doc/benchmarks.md)

[1]: http://tools.ietf.org/rfc/rfc1123.txt


Changelog.md
This is the change log for the main branch of ftpd, which supports
Ruby 1.9 and greater.  For ruby 1.8.7, please use the latest version
before 0.8.0.

### 2.1.0 2017-07-23

* Introduced Release::VERSION and Release::DATE constants
* Fixed all known warnings in tests and example

### 2.0.5 2017-07-23

* Fix DOS caused by fast socket disconnects (thanks iblue)
* Tests pass in Ruby 2.4

### 2.0.4 2017-07-17

* Fix gem description

### 2.0.3 2017-07-17

* Fix erroneous gem release date in gemspec

### 2.0.2 2017-07-17

* Fix high CPU use by bin/ftpdrb (#47)
* ftpdrb now runs from installed gem
* Fix tests hanging in ruby 2.1.7 through 2.1.10	

### 2.0.1 2016-11-02

* Redact password in log (thanks ukolovda)

### 2.0.0 2016-06-01

Breaking changes:

* Freeze strings by default.  This is an unlikely but potential
  breaking change for users of Ruby >= 2.3.

Non-breaking changes

* Added option #nat_ip to configure the advertised IP for passive mode
  connections.

* Added option #passive_ports to configure the port range for passive
  mode server sockets.

### 1.1.1 (2014-11-22)

Administration

* Correct Changelog

### 1.1.0 (2014-11-22)

Features

* Added Server#exception_handler= (thanks akelmanson)
* Added Server#join (thanks akelmanson)	

### 1.0.1

* Less strict gem dependencies

### 1.0.0

Administration
	
* Update gem versions
* Declare stability (v. 1.0.0)

### 0.17.0

This release is brought to you by Mike Ragalie.  Thanks, Mike!

Features

* Streaming file transfers (issues #12 and #26)

API Changes

These breaking API changes are for streaming file transfers.  Custom
file systems will notice these changes.  Nobody else should.

* The FileSystemErrorTranslator has been removed.  Instead, the file
  system may raise FtpServerError with a message and an optional error
  code.

* DiskFileSystem#write now takes an `Ftpd::Stream` object instead of a
  `String`.

* DiskFileSystem#read expects to receive a block and yields an `IO`
  object to the block.

### 0.16.0

Bug fixes

* Fix Errno::ENOTSOCK error in Ruby 2.1.2 (issue #25)

### 0.15.0

Bug fixes

* Ignore Errno::ENOTCONN on socket shutdown (issue #24)

Administration

* Remove examples/foo.rb (accidental commit)

### 0.14.0

Administration

* Remove markdown link (but leave its text) in package description

### 0.13.0

Administration

* Remove badges from package description

### 0.12.0

Bug fixes

* Fix Errno::EADDRINUSE when reusing port (issue #23)	

### 0.11.0

Bug fixes

* Fix Bad file descriptor exception on stop (issue #20)
* CWD returns 250, not 257 (issue #18)

Enhancements

* MDTM command (issue #19)
* SIZE command (issue #19)

### 0.10.0

Bug fixes

* Do not die when implicit SSL connection disconnects (issue #13)	

API Changes:

* Change default interface from "localhost" to "127.0.0.1".

### 0.9.0

Enhancements

* Added example showing ftp used as a test harness with rspec
* Ignore LIST/NLST switches such as "-a"
* Support IPV6

### 0.8.0

Administration

* Split Ruby 1.8 into separate branch
* Upgraded gems

### 0.7.0

Bug fixes

* Gracefully handle Errno::ENOTCONN during socket shutdown (fixes
  gh-1)

### 0.6.0

Enhancements

* Configurable maximum connections (defaults to unlimited).
* Configurable maximum connections per IP (defaults to unlimited).
* Configurable disconnect after too many failed login attempts
  (defaults to unlimited)
* Delay after failed login (configurable).

API Changes

* Changes to {Ftpd::FtpServer} attributes should now only be made
  before calling #start.  The effect of setting these attributes
  after #start is undefined.
* Added {Ftpd::FtpServer#max_connections}
* Added {Ftpd::FtpServer#max_connections_per_ip}
* Added {Ftpd::FtpServer#max_failed_logins}
* Added {Ftpd::FtpServer#failed_login_delay}
* Support FEAT (feature list)
* Support OPTS (set options)

### 0.5.0

Bug fixes

* Replies are sent with the correct line ending ("\r\n" instead of
  "\n")
* Do not hang on out-of-band commands.
* When data connection disconnects, send "426 Connection closed"
  response instead of ending the session.

Enhancements

* Now unconditionally compliant
* Configurable session timeout (see {Ftpd::FtpServer#session_timeout}).
  Defaults to 5 minutes.
* Disable Nagle algorithm on control connection to decrease latency.
  This makes the tests run much faster.
* Support STAT (server status).
* Example has --timeout option for session idle timeout.
* Write log to Logger (see {Ftpd::FtpServer#log}).
* Disallow active-mode connections to privileged ports (configurable).
  See RFC 2577 section 3.
* Added benchmarks.
* Support telnet sequences.

API Changes

* Added {Ftpd::FtpServer#server_name}
* Added {Ftpd::FtpServer#server_version}
* Removed #debug and #debug_path from Ftpd::FtpServer.  They have been
  replaced with #log
# Added {Ftpd::FtpServer#allow_low_data_ports}

### 0.4.0

Enhancements

* Improved driver and file-system documentation.
* Added {Ftpd::ReadOnlyDiskFileSystem read only disk file system}
* Example can be run with a read-only file system
* Supports three different levels of authentication:
  * User
  * User + Password
  * User + Password + Account
* Added --auth switch to the example to select the authentication
  level.
* Support APPE
* Support TYPE "A T" (ASCII/Telnet)
* Support TYPE "LOCAL 8"
* Added switches to example to set authentication values
  * --user
  * --password
  * --account

API changes

* {Example::Driver#authenticate authenticate} now takes three
  parameters (user, password, account).  For compatability, it can be
  defined to take only two, provided you are not doing account
  authentication.
* Added {Ftpd::FtpServer#auth_level} option
* Added {Ftpd::DiskFileSystem::Append}

### 0.3.1

API changes

The file system interface for directory listing was completely
rewritten.  It no longer shells out to ls, which removes potential
command injection security holes, and improves prospects for
portability.

* Removed Ftpd::DiskFileSystem::Ls

FakeFtpServer.rb
#!/usr/bin/env ruby

require 'fileutils'
require 'openssl'
require 'pathname'
require File.expand_path('FakeTlsServer', File.dirname(__FILE__))
require File.expand_path('TempDir', File.dirname(__FILE__))
require File.expand_path('q', File.dirname(__FILE__))

class FakeFtpServer < FakeTlsServer

  attr_accessor :user
  attr_accessor :password
  attr_accessor :debug_path
  attr_accessor :response_delay
  attr_accessor :implicit_tls

  def initialize(data_path)
    super()
    self.user = 'user'
    self.password = 'password'
    self.debug_path = '/dev/stdout'
    @data_path = Pathname.new(data_path)
    @response_delay = 0
    @implicit_tls = false
  end

  def session(socket)
    Session.new(:socket => socket,
                :user => user,
                :password => password,
                :data_path => @data_path,
                :debug_path => debug_path,
                :response_delay => response_delay,
                :implicit_tls => @implicit_tls).run
  end

  private

  class Session

    def initialize(args)
      @socket = args[:socket]
      @socket.encrypt if args[:implicit_tls]
      @expected_user = args[:user]
      @expected_password = args[:password]
      @data_path = @cwd = args[:data_path].realpath
      @debug_path = args[:debug_path]
      @data_type = 'A'
      @mode = 'S'
      @format = 'N'
      @structure = 'F'
      @response_delay = args[:response_delay]
      @data_channel_protection_level = :clear
    end

    def run
      reply "220 FakeFtpServer"
      @state = :user
      catch :done do
        loop do
          begin
            s = get_command
            syntax_error unless s =~ /^(\w+)(?: (.*))?$/
            command, argument = $1.downcase, $2
            unless VALID_COMMANDS.include?(command)
              error "500 Syntax error, command unrecognized: #{s}"
            end
            method = 'cmd_' + command
            unless self.class.private_method_defined?(method)
              error "502 Command not implemented: #{command}"
            end
            send(method, argument)
          rescue Error => e
            reply e.message
          rescue Errno::ECONNRESET, Errno::EPIPE
          end
        end
      end
    end

    private

    class Error < StandardError
    end

    VALID_COMMANDS = [
      "abor",
      "acct",
      "allo",
      "appe",
      "auth",
      "pbsz",
      "cdup",
      "cwd",
      "dele",
      "help",
      "list",
      "mkd",
      "mode",
      "nlst",
      "noop",
      "pass",
      "pasv",
      "port",
      "prot",
      "pwd",
      "quit",
      "rein",
      "rest",
      "retr",
      "rmd",
      "rnfr",
      "rnto",
      "site",
      "smnt",
      "stat",
      "stor",
      "stou",
      "stru",
      "syst",
      "type",
      "user",
    ]

    def cmd_user(argument)
      bad_sequence unless @state == :user
      @user = argument
      @state = :password
      reply "331 Password required"
    end

    def bad_sequence
      error "503 Bad sequence of commands"
    end

    def cmd_pass(argument)
      bad_sequence unless @state == :password
      password = argument
      if @user != @expected_user || password != @expected_password
        @state = :user
        error "530 Login incorrect"
      end
      reply "230 Logged in"
      @state = :logged_in
    end

    def cmd_quit(argument)
      check_logged_in
      reply "221 Byebye"
      @state = :user
    end

    def syntax_error
      error "501 Syntax error"
    end

    def cmd_port(argument)
      check_logged_in
      pieces = argument.split(/,/)
      syntax_error unless pieces.size == 6
      pieces.collect! do |s|
        syntax_error unless s =~ /^\d{1,3}$/
        i = s.to_i
        syntax_error unless (0..255) === i
        i
      end
      @data_hostname = pieces[0..3].join('.')
      @data_port = pieces[4] << 8 | pieces[5]
      reply "200 PORT command successful"
    end

    def cmd_stor(argument)
      close_data_server_socket_when_done do
        check_logged_in
        path = argument
        syntax_error unless path
        target = target_path(path)
        ensure_path_is_in_data_dir(target)
        contents = receive_file(path)
        write_file(target, contents)
        reply "226 Transfer complete"
      end
    end

    def cmd_retr(argument)
      close_data_server_socket_when_done do
        check_logged_in
        path = argument
        syntax_error unless path
        target = target_path(path)
        ensure_path_is_in_data_dir(target)
        contents = read_file(target)
        transmit_file(contents)
      end
    end

    def cmd_dele(argument)
      check_logged_in
      path = argument
      error "501 Path required" unless path
      target = target_path(path)
      ensure_path_is_in_data_dir(target)
      File.unlink(target)
      reply "250 DELE command successful"
    end

    def cmd_list(argument)
      ls(argument, '-l')
    end

    def cmd_nlst(argument)
      ls(argument, '-1')
    end

    def ls(path, option)
      close_data_server_socket_when_done do
        check_logged_in
        ls_dir, ls_path = get_ls_dir_and_path(path)
        list = get_file_list(ls_dir, ls_path, option)
        transmit_file(list, 'A')
      end
    end

    def get_ls_dir_and_path(path)
      path = path || '.'
      target = target_path(path)
      target = realpath(target)
      ensure_path_is_in_data_dir(target)
      if target.to_s.index(@cwd.to_s) == 0
        ls_dir = @cwd
        ls_path = target.to_s[@cwd.to_s.length..-1]
      else
        raise
      end
      if ls_path =~ /^\//
        ls_path = $'
      end
      [ls_dir, ls_path]
    end

    def get_file_list(ls_dir, ls_path, option)
      command = [
        'ls',
        option,
        ls_path,
        '2>&1',
      ].compact.join(' ')
      list = Dir.chdir(ls_dir) do
        `#{command}`
      end
      list = "" if $? != 0
      list = list.gsub(/^total \d+\n/, '')
      list
    end

    def realpath(pathname)
      handle_system_error do
        basename = File.basename(pathname.to_s)
        if is_glob?(basename)
          pathname.dirname.realpath + basename
        else
          pathname.realpath
        end
      end
    end

    def is_glob?(filename)
      filename =~ /[.*]/
    end

    def cmd_type(argument)
      check_logged_in
      syntax_error unless argument =~ /^(\S)(?: (\S+))?$/
      type_code = $1
      format_code = $2
      set_type(type_code)
      set_format(format_code)
      reply "200 Type set to #{@data_type}"
    end

    def set_type(type_code)
      name, implemented = DATA_TYPES[type_code]
      error "504 Invalid type code" unless name
      error "504 Type not implemented" unless implemented
      @data_type = type_code
    end

    def set_format(format_code)
      format_code ||= 'N'
      name, implemented = FORMAT_TYPES[format_code]
      error "504 Invalid format code" unless name
      error "504 Format not implemented" unless implemented
      @data_format = format_code
    end

    def cmd_mode(argument)
      check_logged_in
      name, implemented = TRANSMISSION_MODES[argument]
      error "504 Invalid mode code" unless name
      error "504 Mode not implemented" unless implemented
      @mode = argument
      reply "200 Mode set to #{name}"
    end

    def cmd_stru(argument)
      check_logged_in
      name, implemented = FILE_STRUCTURES[argument]
      error "504 Invalid structure code" unless name
      error "504 Structure not implemented" unless implemented
      @structure = argument
      reply "200 File structure set to #{name}"
    end

    def cmd_noop(argument)
      reply "200 Nothing done"
    end

    def cmd_pasv(argument)
      check_logged_in
      if @data_server
        reply "200 Already in passive mode"
      else
        @data_server = TCPServer.new('localhost', 0)
        ip = @data_server.addr[3]
        port = @data_server.addr[1]
        quads = [
          ip.scan(/\d+/),
          port >> 8,
          port & 0xff,
        ].flatten.join(',')
        reply "227 Entering passive mode (#{quads})"
      end
    end

    def cmd_cwd(argument)
      check_logged_in
      target = if argument =~ %r"^/(.*)$"
                 @data_path + $1
               else
                 @cwd + argument
               end
      ensure_path_is_in_data_dir(target)
      restore_cwd_on_error do
        @cwd = target
        pwd
      end
    end

    def cmd_cdup(argument)
      check_logged_in
      cmd_cwd('..')
    end

    def cmd_pwd(argument)
      check_logged_in
      pwd
    end

    def cmd_auth(security_scheme)
      if @socket.encrypted?
        raise Error, "503 AUTH already done"
      end
      unless security_scheme =~ /^TLS(-C)?$/i
        raise Error, "500 Security scheme not implemented: #{security_scheme}"
      end
      reply "234 AUTH #{security_scheme} OK."
      @socket.encrypt
    end

    def cmd_pbsz(buffer_size)
      syntax_error unless buffer_size =~ /^\d+$/
      buffer_size = buffer_size.to_i
      unless @socket.encrypted?
        raise Error, "503 PBSZ must be preceded by AUTH"
      end
      unless buffer_size == 0
        raise Error, "501 PBSZ=0"
      end
      reply "200 PBSZ=0"
      @protection_buffer_size_set = true
    end

    def cmd_prot(level_arg)
      level_code = level_arg.upcase
      unless @protection_buffer_size_set
        raise Error, "503 PROT must be preceded by PBSZ"
      end
      level = DATA_CHANNEL_PROTECTION_LEVELS[level_code]
      unless level
        raise Error, "504 Unknown protection level"
      end
      unless level == :private
        raise Error, "536 Unsupported protection level #{level}"
      end
      @data_channel_protection_level = level
      reply "200 Data protection level #{level_code}"
    end

    def pwd
      reply "250 OK. Current directory is #{sanitized_cwd}"
    end

    def relative_to_data_path(path)
      data_path = realpath(@data_path).to_s
      path = realpath(path).to_s
      path = path.gsub(data_path, '')
      path = '/' if path.empty?
      path
    end

    def sanitized_cwd
      relative_to_data_path(@cwd)
    end

    def error(message)
      raise Error, message
    end

    TRANSMISSION_MODES = {
      'B'=>['Block', false],
      'C'=>['Compressed', false],
      'S'=>['Stream', true],
    }

    FORMAT_TYPES = {
      'N'=>['Non-print', true],
      'T'=>['Telnet format effectors', false],
      'C'=>['Carriage Control (ASA)', false],
      'L'=>['Local byte size', false],
    }

    DATA_TYPES = {
      'A'=>['ASCII', true],
      'E'=>['EBCDIC', false],
      'I'=>['BINARY', true],
      'L'=>['LOCAL', false],
    }

    FILE_STRUCTURES = {
      'R'=>['Record', false],
      'F'=>['File', true],
      'P'=>['Page', false],
    }

    DATA_CHANNEL_PROTECTION_LEVELS = {
      'C'=>:clear,
      'S'=>:safe,
      'E'=>:confidential,
      'P'=>:private
    }

    def check_logged_in
      return if @state == :logged_in
      error "530 Not logged in"
    end

    def ensure_path_is_in_data_dir(path)
      unless child_path_of?(@data_path, path)
        error "550 Access denied"
      end
    end

    def child_path_of?(parent, child)
      child.cleanpath.to_s.index(parent.cleanpath.to_s) == 0
    end

    def target_path(path)
      path = Pathname.new(path)
      base, path = if path.to_s =~ /^\/(.*)/
                     [@data_path, $1]
                   else
                     [@cwd, path]
                   end
      base + path
    end

    def read_file(path)
      handle_system_error do
        File.open(path, 'rb') do |file|
          file.read
        end
      end
    end

    def write_file(dest, contents)
      handle_system_error do
        File.open(dest, 'w') do |file|
          file.write(contents)
        end
      end
    end

    def handle_system_error
      begin
        yield
      rescue SystemCallError => e
        error "550 #{e}"
      end
    end

    def transmit_file(contents, data_type = @data_type)
      open_data_connection do |data_socket|
        contents = native_to_ascii(contents) if data_type == 'A'
        data_socket.write(contents)
        debug("Sent #{contents.size} bytes")
        reply "226 Transfer complete"
      end
    end

    def receive_file(path)
      open_data_connection do |data_socket|
        contents = data_socket.read
        contents = ascii_to_native(contents) if @data_type == 'A'
        debug("Received #{contents.size} bytes")
        contents
      end
    end

    def native_to_ascii(s)
      s.gsub(/\n/, "\r\n")
    end

    def ascii_to_native(s)
      s.gsub(/\r\n/, "\n")
    end

    def open_data_connection(&block)
      reply "150 Opening #{data_connection_description}"
      if @data_server
        if encrypt_data?
          open_passive_tls_data_connection(&block)
        else
          open_passive_data_connection(&block)
        end
      else
        if encrypt_data?
          open_active_tls_data_connection(&block)
        else
          open_active_data_connection(&block)
        end
      end
    end

    def data_connection_description
      [
        DATA_TYPES[@data_type][0],
        "mode data connection",
        ("(TLS)" if encrypt_data?)
      ].compact.join(' ')
    end

    def encrypt_data?
      @data_channel_protection_level != :clear
    end

    def open_active_data_connection
      data_socket = TCPSocket.new(@data_hostname, @data_port)
      begin
        yield(data_socket)
      ensure
        data_socket.close
      end
    end

    def open_active_tls_data_connection
      open_active_data_connection do |socket|
        make_tls_connection(socket) do |ssl_socket|
          yield(ssl_socket)
        end
      end
    end

    def open_passive_data_connection
      data_socket = @data_server.accept
      begin
        yield(data_socket)
      ensure
        data_socket.close
      end
    end

    def close_data_server_socket_when_done
      yield
    ensure
      close_data_server_socket
    end

    def close_data_server_socket
      return unless @data_server
      @data_server.close
      @data_server = nil
    end

    def open_passive_tls_data_connection
      open_passive_data_connection do |socket|
        make_tls_connection(socket) do |ssl_socket|
          yield(ssl_socket)
        end
      end
    end

    def make_tls_connection(socket)
      ssl_socket = OpenSSL::SSL::SSLSocket.new(socket, @socket.ssl_context)
      ssl_socket.accept
      begin
        yield(ssl_socket)
      ensure
        ssl_socket.close
      end
    end

    def get_command
      s = @socket.gets
      throw :done if s.nil?
      s = s.chomp
      debug(s)
      s
    end

    def reply(s)
      if @response_delay.to_i != 0
        debug "#{@response_delay} second delay before replying"
        sleep @response_delay
      end
      debug(s)
      @socket.puts(s)
    end

    def debug(*s)
      return unless debug?
      File.open(@debug_path, 'a') do |file|
        file.puts(*s)
      end
    end

    def debug?
      ENV['DEBUG'].to_i != 0
    end

    def restore_cwd_on_error
      orig_cwd = @cwd
      yield
    rescue
      @cwd = orig_cwd
      raise
    end

  end

end

class Scaffold

  def initialize
    @data_dir = TempDir.new
    create_files
    @server = FakeFtpServer.new(@data_dir.path)
    set_credentials
    display_connection_info
    create_connection_script
  end

  def run
    wait_until_stopped
  end

  private
  
  HOST = 'localhost'

  def create_files
    [
      'README',
      'outgoing/getme',
    ].each do |path|
      base_name = File.basename(path)
      dir_name = File.dirname(path)
      dir_path = File.join(@data_dir.path, dir_name)
      file_path = File.join(dir_path, base_name)
      FileUtils.mkdir_p(dir_path)
      File.open(file_path, 'w') do |file|
        file.puts "Contents of #{path}"
      end
    end
  end

  def set_credentials
    @server.user = ENV['LOGNAME']
    @server.password = ''
  end

  def display_connection_info
    puts "Host: #{HOST}"
    puts "Port: #{@server.port}"
    puts "User: #{@server.user}"
    puts "Pass: #{@server.password}"
    puts "Directory: #{@data_dir.path}"
  end

  def create_connection_script
    command_path = '/tmp/connect_to_fake_ftp_server.sh'
    File.open(command_path, 'w') do |file|
      file.puts "#!/bin/bash"
      file.puts "ftp $FTP_ARGS #{HOST} #{@server.port}"
    end
    system("chmod +x #{command_path}")
    puts "Connection script written to #{command_path}"
  end

  def wait_until_stopped
    puts "FTP server started.  Press ENTER or c-C to stop it"
    $stdout.flush
    begin
      gets
    rescue Interrupt
      puts "Interrupt"
    end
  end

end

Scaffold.new.run if $0 == __FILE__
